moengage-segment-kotlin-destination/src/main/java/com/segment/analytics/kotlin/destinations/moengage/MoEngageDestination.kt

```diff
@@ -0,0 +1,242 @@
+                            city,
+                            instanceId
+                        )
+                    }
+                    val country = address.jsonObject.getString(USER_TRAIT_ADDRESS_COUNTRY)
+                    if (!country.isNullOrEmpty()) {
+                        MoEAnalyticsHelper.setUserAttribute(
+                            application.applicationContext,
+                            USER_TRAIT_ADDRESS_COUNTRY,
+                            country,
+                            instanceId
+                        )
+                    }
+                    val state = address.jsonObject.getString(USER_TRAIT_ADDRESS_STATE)
+                    if (!state.isNullOrEmpty()) {
+                        MoEAnalyticsHelper.setUserAttribute(
+                            application.applicationContext,
+                            USER_TRAIT_ADDRESS_STATE,
+                            state,
+                            instanceId
+                        )
+                    }
+                }
+            }
+            val location = payload.traits[USER_TRAIT_LOCATION]
+            if (location != null && location.jsonObject.isNotEmpty()) {
+                MoEAnalyticsHelper.setUserAttribute(
+                    application.applicationContext, USER_ATTRIBUTE_USER_LOCATION, GeoLocation(
+                        location.jsonObject.getDouble(USER_TRAIT_LOCATION_LATITUDE) ?: 0.0,
+                        location.jsonObject.getDouble(USER_TRAIT_LOCATION_LONGITUDE) ?: 0.0
+                    ), instanceId
+                )
+            }
+        } catch (t: Throwable) {
+            Logger.print(LogLevel.ERROR, t) { "$tag identify(): " }
+        }
+
+        return payload
+    }
+
+    override fun reset() {
+        try {
+            super.reset()
+            Logger.print { "$tag reset(): will try to reset" }
+            MoECoreHelper.logoutUser(application.applicationContext, instanceId)
+        } catch (t: Throwable) {
+            Logger.print(LogLevel.ERROR, t) { "$tag reset(): " }
+        }
+    }
+
+    override fun track(payload: TrackEvent): BaseEvent {
+        try {
+            super.track(payload)
+            Logger.print { "$tag track(): will try to track TrackEvent" }
+            if (payload.properties.isNotEmpty()) {
+                integrationHelper.trackEvent(
+                    payload.event,
+                    payload.properties.toJSONObject(),
+                    instanceId
+                )
+            } else {
+                integrationHelper.trackEvent(payload.event, JSONObject(), instanceId)
+            }
+        } catch (t: Throwable) {
+            Logger.print(LogLevel.ERROR, t) { "$tag track(): " }
+        }
+        return payload
+    }
+
+    override fun version(): String {
+        return BuildConfig.MOENGAGE_SEGMENT_KOTLIN_VERSION
+    }
+
+    private fun trackAnonymousId() {
+        Executors.newSingleThreadExecutor().submit {
```

Comment:
Surround thread creation with `try-catch` to avoid memory errors leading to crashes.

Reply:
Done